File: lintcode/python/0364_trapping_rain_water_II.py
#!/usr/bin/python -t

# heap
# 将矩阵周边的格子都放到堆里，这些格子上面是无法盛水的。
# 每次在堆里挑出一个高度最小的格子 cell，把周围的格子加入到堆里。
# 这些格子被加入堆的时候，计算他们上面的盛水量。
# 
# 盛水量 = cell.height - 这个格子的高度
# 当然如果这个值是负数，盛水量就等于 0。

# O  (  n*m      +         m*n  * log(2n+2m))
#     加入heap              遍历   堆操作

import heapq

class Solution:
    """
    @param heights: a matrix of integers
    @return: an integer
    """
    def trapRainWater(self, heights):
        # write your code here
        m = len(heights)
        if m == 0:
            return 0
        n = len(heights[0])
        if n == 0:
            return 0
            
        self.m = m
        self.n = n
        ret = 0
            
        self.build_border(heights)
        
        while len(self.hq) != 0:
            height, x, y = heapq.heappop(self.hq)
            
            for nx, ny in self.adjacents(x, y):
                ret += max(0, height - heights[nx][ny])
                self.add_border(nx, ny, max(heights[nx][ny], height))
                
        return ret
                
            
    
    def build_border(self, heights):
        self.hq = []
        self.v = set()
        
        for i in range(self.m):
            self.add_border(i, 0, heights[i][0])
            self.add_border(i, self.n-1, heights[i][self.n-1])
            
        for j in range(self.n):
            self.add_border(0, j, heights[0][j])
            self.add_border(self.m-1, j, heights[self.m-1][j])
            
        return
        
        
    def add_border(self, x, y, height):
        heapq.heappush(self.hq, (height, x, y))
        self.v.add((x,y))
        
        return
    
        
    def adjacents(self, x, y):
        neis = []
        dx = [1, -1, 0, 0]
        dy = [0, 0, 1, -1]
        
        for i in range(4):
            cx = x + dx[i]
            cy = y + dy[i]
            
            if 0 <= cx < self.m and 0 <= cy < self.n and (cx, cy) not in self.v:
                neis.append((cx, cy))
                
        return neis
        

File: lintcode/golang/0364_trapping_rain_water_II.go
package main

// heap time mnlog(m+n)

import (
	"container/heap"
	"fmt"
)

type locate struct {
	x int
	y int
}

type Node struct {
	val int
	x   int
	y   int
}

// An IntHeap is a min-heap of ints.
type NodeHeap []*Node

func (h NodeHeap) Len() int           { return len(h) }
func (h NodeHeap) Less(i, j int) bool { return h[i].val < h[j].val }
func (h NodeHeap) Swap(i, j int)      { h[i], h[j] = h[j], h[i] }

func (h *NodeHeap) Push(x interface{}) {
	// Push and Pop use pointer receivers because they modify the slice's length,
	// not just its contents.
	*h = append(*h, x.(*Node))
}

func (h *NodeHeap) Pop() interface{} {
	old := *h
	n := len(old)
	x := old[n-1]
	*h = old[0 : n-1]
	return x
}

/**
 * @param heights: a matrix of integers
 * @return: an integer
 */
func trapRainWater(heights [][]int) int {
	// write your code here
	minHp := &NodeHeap{}
	v := make(map[locate]bool)
	m := len(heights)
	n := len(heights[0])
	var ret int

	for i := 0; i < m; i++ {
		for j := 0; j < n; j++ {
			if i == 0 || i == m-1 || j == 0 || j == n-1 {
				heap.Push(minHp, &Node{heights[i][j], i, j})
				v[locate{i, j}] = true
			}
		}
	}

	dir := []struct {
		dx int
		dy int
	}{{1, 0}, {-1, 0}, {0, 1}, {0, -1}}

	for minHp.Len() != 0 {
		node := heap.Pop(minHp).(*Node)
		for i := 0; i < len(dir); i++ {
			cx := dir[i].dx + node.x
			cy := dir[i].dy + node.y
			if cx >= 0 && cx < m && cy >= 0 && cy < n && v[locate{cx, cy}] == false {
				maxHeight := max(node.val, heights[cx][cy])
				ret += maxHeight - heights[cx][cy]
				heap.Push(minHp, &Node{maxHeight, cx, cy})
				v[locate{cx, cy}] = true

			}

		}
	}

	return ret
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

// This example inserts several ints into an IntHeap, checks the minimum,
// and removes them in order of priority.
func main() {
	a := [][]int{
		{12, 13, 0, 12},
		{13, 4, 13, 12},
		{13, 8, 10, 12},
		{12, 13, 12, 12},
		{13, 13, 13, 13},
	}
	fmt.Println(trapRainWater(a))
}



import (
	"container/heap"
	"fmt"
)

type Node struct {
	val int
	x   int
	y   int
}

type MinHeap []*Node

func (h MinHeap) Len() int           { return len(h) }
func (h MinHeap) Less(i, j int) bool { return h[i].val < h[j].val }
func (h MinHeap) Swap(i, j int)      { h[i], h[j] = h[j], h[i] }

func (h *MinHeap) Push(x interface{}) {
	*h = append(*h, x.(*Node))
}

func (h *MinHeap) Pop() interface{} {
	cur := *h
	n := len(cur)
	x := cur[n-1]
	*h = cur[:n-1]
	return x
}

/**
 * @param heights: a matrix of integers
 * @return: an integer
 */
func trapRainWater(heights [][]int) int {
	// write your code here
	m := len(heights)
	if m == 0 {
		return 0
	}
	n := len(heights[0])
	if n == 0 {
		return 0
	}
	minH := &MinHeap{}
	v := make([][]bool, m)
	for i := range v {
		v[i] = make([]bool, n)
	}

	for i := 0; i < m; i++ {
		for j := 0; j < n; j++ {
			if i == 0 || i == m-1 || j == 0 || j == n-1 {
				heap.Push(minH, &Node{val: heights[i][j], x: i, y: j})
				v[i][j] = true
			}
		}
	}

	dir := []struct {
		dx int
		dy int
	}{
		{1, 0},
		{-1, 0},
		{0, 1},
		{0, -1},
	}

	ret := 0

	for minH.Len() != 0 {
		node := heap.Pop(minH).(*Node)
		for i := 0; i < len(dir); i++ {
			nx := dir[i].dx + node.x
			ny := dir[i].dy + node.y
			if nx >= 0 && nx < m && ny >= 0 && ny < n && !v[nx][ny] {
				maxH := max(node.val, heights[nx][ny])
				ret += maxH - heights[nx][ny]
				heap.Push(minH, &Node{val: maxH, x: nx, y: ny})
				v[nx][ny] = true
			}
		}
	}

	return ret
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}
